File: csharp/dotnet/pxprpc/backend/TcpBackend.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace pxprpc.backend
{
    public class TcpWorkThread
    {

        public TcpClient s;
        public ServerContext sc;
        public TcpBackend attached;
        public TcpWorkThread(TcpClient s, ServerContext sc)
        {
            this.s = s;
            this.sc = sc;
        }
        public void run()
        {
            try
            {
                sc.init(s.GetStream());
                sc.serve();
            }
            catch (Exception e)
            {
                //catch all exception to avoid breaking Application unexpectedly.
                if (!(e is IOException))
                {
                    Console.Error.Write(e.Message);
                    Console.Error.Write(e.StackTrace);
                }
            }
            finally
            {
                sc.Dispose();
                if (attached != null)
                {
                    attached.runningContext.Remove(this);
                }
            }

        }
    }
    public class TcpBackend : IDisposable
    {
        public TcpBackend()
        {
        }
        public String bindAddr;
        public Dictionary<String, Object> funcMap = new Dictionary<String, Object>();
        protected TcpListener ss;
        public HashSet<TcpWorkThread> runningContext = new HashSet<TcpWorkThread>();

        protected void execute(Action r)
        {
            Thread t = new Thread(() => r());
            t.IsBackground = true;
            t.Start();
        }
        protected bool running = false;
        public void listenAndServe()
        {
            ss = new TcpListener(IPEndPoint.Parse(bindAddr));
            ss.Start();
            this.running = true;
            while (running)
            {
                var soc = ss.AcceptTcpClient();
                ServerContext sc = new ServerContext();
                sc.funcMap = this.funcMap;
                TcpWorkThread wt = new TcpWorkThread(soc, sc);
                wt.attached = this;
                runningContext.Add(wt);
                execute(wt.run);

            }
        }
        public void Dispose()
        {
            //Copy runningContext to avoid modifing to iterating
            this.running = false;
            List<TcpWorkThread> snapshot = new List<TcpWorkThread>();
            snapshot.AddRange(runningContext);
            foreach (TcpWorkThread wt in snapshot)
            {
                if (wt.sc.running)
                {
                    wt.sc.Dispose();
                }
            }
            ss.Stop();
        }
    }
}

File: java/src/pxprpc/backend/TCPBackend.java
package pxprpc.backend;

import java.io.Closeable;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.channels.AsynchronousCloseException;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

import pxprpc.base.FuncMap;
import pxprpc.base.ServerContext;

public class TCPBackend implements Closeable{
	/* Known issue. When use ChannelIo2, 'write' may blocked until 'read' return on android lower than 7.0.
	May also occur on java version lower than 1.8.
	So we use ChannelIo(1) until ChannelIo2 is stable enough with significant performance advantage. */
	public static int channelIoVersion=1;

	public TCPBackend() {
	}
	public InetSocketAddress bindAddr;
	public FuncMap funcMap;
	protected ServerSocketChannel ss;
	public Set<WorkThread> runningContext=new HashSet<WorkThread>();
	public static class WorkThread implements Runnable{
		public SocketChannel s;
		public ServerContext sc;
		public TCPBackend attached;
		public WorkThread(SocketChannel s,ServerContext sc) {
			this.s=s;
			this.sc=sc;
		}

		@Override
		public void run() {
			try {
				if(TCPBackend.channelIoVersion==1){
					sc.init(new ChannelIo(s), attached.funcMap);
				}else{
					sc.init(new ChannelIo2(s,s), attached.funcMap);
				}
				s.configureBlocking(true);
				sc.serve();
			} catch (Exception e) {
				//catch all exception to avoid breaking Application unexpectedly.
				if(!(e instanceof IOException)){
					e.printStackTrace();
				}
			}finally{
				try {sc.close();} catch (IOException e) {}
				if(attached!=null) {
					attached.runningContext.remove(this);
				}
			}
			
		}
	}
	protected void execute(Runnable r) {
		Thread t=new Thread(r);
		t.setDaemon(true);
		t.start();
	}
	public void listenAndServe() throws IOException {
		try {
			ss=ServerSocketChannel.open();
			//chan.bind since 1.7, so we choose socket.bind.
			ss.socket().bind(bindAddr);
			while(ss.isOpen()) {
				SocketChannel soc = ss.accept();
				ServerContext sc = new ServerContext();
				WorkThread wt = new WorkThread(soc, sc);
				wt.attached=this;
				runningContext.add(wt);
				execute(wt);
			}
		}catch(AsynchronousCloseException c) {}
	}
	
	@Override
	public void close() throws IOException {
		//Copy runningContext to avoid modifing to iterating
		ArrayList<WorkThread> snapshot = new ArrayList<WorkThread>();
		snapshot.addAll(runningContext);
		for(WorkThread wt:snapshot) {
			if(wt.sc.running) {
				wt.sc.close();
			}
		}
		ss.close();
	}
}
